### PR TITLE
Add template revision button

### LINKED ISSUE #44503: Template inspector: Surface revisions panel
Templates and template parts [support revisions](https://github.com/WordPress/wordpress-develop/blob/b4acf193d1cac4b287492b762dfc2344b0942cbe/src/wp-includes/post.php#L388). We should surface these in the inspector for Templates and Template Parts, especially because those inspectors are already so empty.

It could be surfaced in the same way as it is in the post editor, just as a link to the revisions page:

Revisions in template inspector, just a link

### PR DESCRIPTION
**Pull request by georgeh**
## What?
Adds a button to templates and template parts to get to the revision compare screen

## Why?

Closes #44503

## How?

This adds a button to the site editor for Templates and Template Parts that links to the revision viewer for that template.

It adds 2 new fields to the REST response for `/wp/v2/template/{theme}//{part}` and `/wp/v2/template-part/{theme}//{part}` routes: 'revision_count' and `latest_revision_id`. The Posts REST controller uses the `_links` section to expose these and links to routes for the Revisions REST controller. I chose to add these to the body because the Revisions REST routes assume numeric IDs with `/revisions` at the end, and the existing template routes receive everything in their namespace. Adding `/revisions` routes for templates and template-parts would require complex RegExes that could potentially lead to a DoS attack. However if someone can get the routes on the Revisions REST controller to work with Template IDs, I would be happy to add them.

## Testing Instructions

1. Open the Site Editor
2. Edit a Template or Template Part
3. Save to create a custom template
4. Make another change and save again, to create a second custom template
5. Confirm that the revision button appears in the Template sidebar and links to the revision viewer

## Screenshots or screencast 

https://user-images.githubusercontent.com/107534/217497598-9b112429-2a03-4fc2-b2e2-c1a8537acf21.mp4

### CODE REVIEW

**Comment by georgeh** [comment]
Thinking more about the REST routes in this, I think I can use the revisions REST controller, but only add the links when there are 2 or more custom versions of a template.

As a follow-up it would also be good to let the revision viewer load theme templates from the filesystem as a kind of revision 0

**Comment by georgeh** [comment]
This button could probably use some design help too:

image

**Comment by georgeh** [comment]
> Thinking more about the REST routes in this, I think I can use the revisions REST controller, but only add the links when there are 2 or more custom versions of a template.

No, it won't work with the routes as they are currently set up. I've learned this twice now so I'm writing it down here for the next time I attempt this.

`WP_REST_Templates_Controller` registers a route for
```php
// The route.
sprintf(
	'/%s/(?P<id>%s%s)',
	$this->rest_base,
	// Matches theme's directory: `/themes/<subdirectory>/<theme>/` or `/themes/<theme>/`.
	// Excludes invalid directory name characters: `/:<>*?"|`.
	'([^\/:<>\*\?"\|]+(?:\/[^\/:<>\*\?"\|]+)?)',
	// Matches the template name.
	'[\/\w-]+'
)
```
which expands to something like `/wp/v2/templates/twentytwentytwo//footer`

`WP_REST_Revisions_Controller` registers a route for every post type that supports revisions for
```php
// $this->rest_base = 'revisions'
'/' . $this->parent_base . '/(?P<parent>[\d]+)/' . $this->rest_base;
// ... and ....
'/' . $this->parent_base . '/(?P<parent>[\d]+)/' . $this->rest_base . '/(?P<id>[\d]+)',
```
which expands to something like `/wp/v2/templates/123/revisions` and `/wp/v2/templates/123/revisions/456`

Excluding routes that end in `/revisions[/\d]*` is difficult and has the potential to introduce a [ReDoS](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS) vulnerability. Still, if someone has suggestions on how to make the routes on both of those classes compatible, I would love to use the same response format as posts.

File: lib/experimental/class-gutenberg-rest-template-revision-count.php
Comment (spacedmonkey) [inline lib/experimental/class-gutenberg-rest-template-revision-count.php]: Instead of adding the data like this, please add like other endpoints, like [this](https://github.com/WordPress/wordpress-develop/blob/2517d5452561b1171be01ea7d05c1cc1cd68fe9b/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2065-L2081).

File: lib/experimental/rest-api.php
Comment (spacedmonkey) [inline lib/experimental/rest-api.php]: Let's register the endpoints like [this](https://github.com/WordPress/gutenberg/blob/56ec52e296a1381e4cdb2c2b9ef9b03ebe5d7b06/lib/rest-api.php#L105-L122)

File: lib/experimental/class-gutenberg-rest-template-revision-count.php
Comment (spacedmonkey) [inline lib/experimental/class-gutenberg-rest-template-revision-count.php]:
```suggestion
		if ( ! empty( $template->wp_id ) && $template->wp_id > 0 ) {
```

File: lib/experimental/class-gutenberg-rest-template-revision-count.php
Comment (spacedmonkey) [inline lib/experimental/class-gutenberg-rest-template-revision-count.php]: Not that we should include these, as this should be a link. But there should have only be edit. embed.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Good start but changes required.

**Comment by georgeh** [comment]
Working on the revisions, thanks for the review @spacedmonkey. Also I opened #45290 as a possible follow-up to make template revisions clearer

File: lib/experimental/class-gutenberg-rest-template-revision-count.php
Comment (georgeh) [reply lib/experimental/class-gutenberg-rest-template-revision-count.php]: The other endpoints provide links to REST routes that provide revision information, which I inferred as the reason that they are in the `_links` section in the response. I've noted in the How section of my description and [this comment](https://github.com/WordPress/gutenberg/pull/45215#issuecomment-1290942806) that the REST routes for revisions are not compatible with the routes for templates, because template IDs are non-numeric and use slashes internally.

I chose not to put these in the `_links` section because of the difficulty getting the Revision REST controller routes to respond. Are you saying you'd still like to have these data in `_links` without URLs for other REST responses?

Comment (spacedmonkey) [reply lib/experimental/class-gutenberg-rest-template-revision-count.php]: No, I would like to the revisions endpoints working.

Comment (spacedmonkey) [reply lib/experimental/class-gutenberg-rest-template-revision-count.php]: There is no such idea of revisioned templates for file based templates. There will only ever work with ones stored in the database. As we check to see if there are WordPress id based templates. So this is really an issue.

Comment (georgeh) [reply lib/experimental/class-gutenberg-rest-template-revision-count.php]: What do you think about having a different URL structure for templates? Something like `/wp/v2/revisions/template[_part]s/123/456`? It seems like the revision controller could have a special case for the route for those post types to avoid conflicts.

**Comment by priethor** [comment]
Does this PR provide an alternative to https://github.com/WordPress/wordpress-develop/pull/3476, or does it depend on it?

File: lib/experimental/class-gutenberg-rest-template-revision-count.php
Comment (spacedmonkey) [reply lib/experimental/class-gutenberg-rest-template-revision-count.php]: There is already an existing endpoint that looks like this. 

```
/wp/v2/templates/(?P<parent>[\\d]+)/revisions":
```

and 

```
/wp/v2/templates/(?P<parent>[\\d]+)/revisions/(?P<id>[\\d]+)":
```

There is no need for a  different endpoint / structure.

**Comment by georgeh** [comment]
> Does this PR provide an alternative to [WordPress/wordpress-develop#3476](https://github.com/WordPress/wordpress-develop/pull/3476), or does it depend on it?

@priethor That was an earlier attempt on adding to the REST API and this is a better place for that change. I'll close that and we can merge the changes to core using the usual Gutenberg merge process.

File: lib/experimental/class-gutenberg-rest-template-revision-count.php
Comment (georgeh) [reply lib/experimental/class-gutenberg-rest-template-revision-count.php]: Those endpoints do exist, however the routes in `WP_REST_Templates_Controller` also match those URLs and so `WP_REST_Revisions_Controller` never sees requests for them. You can confirm this by making a request to those endpoints and seeing that the response is from the Templates controller: the error code is `rest_template_not_found`.

Here is the best solution I've been able to come up with to route `/wp/v2/template-parts/18/revisions` and `/wp/v2/template-parts/123/revisions/456` to `WP_REST_Revisions_Controller`:

First, change [the route for `WP_REST_Templates_Controller::get_item`](https://github.com/WordPress/WordPress/blob/72ae8d87ac7b8feb864bda854b7a645c915d530d/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L101-L109) to add `(?<!revisions)` at the end. This prevents the controller from matching requests for `/wp/v2/template-parts/123/revisions`. This fixes requests for `/wp/v2/template-parts/18/revisions`.

This regex will still attempt to match `/wp/v2/template-parts/123/revisions/456` because that is indistinguishable from `/wp/v2/template-parts/themedir/twentytwentythree/footer` (we need to [allow for theme dirs](https://github.com/WordPress/WordPress/blob/72ae8d87ac7b8feb864bda854b7a645c915d530d/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L104), and we have to accept [a single slash between theme name and part](https://github.com/WordPress/WordPress/blob/72ae8d87ac7b8feb864bda854b7a645c915d530d/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L189-L192)). `(?<!revisions)` is a negative lookahead, and they must be fixed width so we can't add `\d+` in there to match one-or-more digits.

We may be able to update the regex for the template routes to require at least one non-numeric character in theme and template names **if** we can guarantee no one is going to name a theme `2023` or a template part `1`. If not, I put together this pseudocode that feels incredibly hacky, and not a solution I like, but it should allow us to have `WP_REST_Templates_Controller` intercept requests for `/wp/v2/template-parts/123/revisions/456`:

```php
function wp_api_reroute_template_revision_requests( $dispatch_result, $request, $route, $handler ) {
	// Check if the request is for a template revision.
	if ( is_a( $handler['callback'][0], 'WP_REST_Templates_Controller', true ) ) {
		$post_type        = $handler['callback'][0]->getPostType();
		$post_type_object = get_post_type_object( $post_type );
		$post_base        = ! empty( $post_type_object->rest_base ) ? $post_type_object->rest_base : $post_type_object->name;

		// See if we have a request for get_item()
		if ( preg_match( '@/' . $post_base . '/(?P<parent>[\d]+)/revisions/(?P<id>[\d]+)$@',
		                 $request->get_route(),
						 $get_items_matches ) ) {
			$revision_controller = new WP_REST_Revisions_Controller( $post_type );
			return $revision_controller->get_item( $get_items_matches );
		}
	}
}
add_filter( 'rest_dispatch_request', 'wp_api_reroute_template_revision_requests', 10, 4 );
```

This feels very brittle and indirect, and not a particularly elegant solution, but it is a solution nonetheless.

@spacedmonkey I would appreciate your thoughts on the proposed solutions (don't link to revisions, change the URL structure for template revisions, intercept requests for revisions in a filter), and if you have any other ideas on ensuring that the Revision controller gets the requests instead of the Template controller. Thanks again for your work on this.

**Comment by spacedmonkey** [comment]
> > Thinking more about the REST routes in this, I think I can use the revisions REST controller, but only add the links when there are 2 or more custom versions of a template.
> 
> No, it won't work with the routes as they are currently set up. I've learned this twice now so I'm writing it down here for the next time I attempt this.
> 
> `WP_REST_Templates_Controller` registers a route for
> 
> ```
> // The route.
> sprintf(
> 	'/%s/(?P<id>%s%s)',
> 	$this->rest_base,
> 	// Matches theme's directory: `/themes/<subdirectory>/<theme>/` or `/themes/<theme>/`.
> 	// Excludes invalid directory name characters: `/:<>*?"|`.
> 	'([^\/:<>\*\?"\|]+(?:\/[^\/:<>\*\?"\|]+)?)',
> 	// Matches the template name.
> 	'[\/\w-]+'
> )
> ```
> 
> which expands to something like `/wp/v2/templates/twentytwentytwo//footer`
> 
> `WP_REST_Revisions_Controller` registers a route for every post type that supports revisions for
> 
> ```
> // $this->rest_base = 'revisions'
> '/' . $this->parent_base . '/(?P<parent>[\d]+)/' . $this->rest_base;
> // ... and ....
> '/' . $this->parent_base . '/(?P<parent>[\d]+)/' . $this->rest_base . '/(?P<id>[\d]+)',
> ```
> 
> which expands to something like `/wp/v2/templates/123/revisions` and `/wp/v2/templates/123/revisions/456`
> 
> Excluding routes that end in `/revisions[/\d]*` is difficult and has the potential to introduce a [ReDoS](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS) vulnerability. Still, if someone has suggestions on how to make the routes on both of those classes compatible, I would love to use the same response format as posts.

I have reported this issue upstream here - https://core.trac.wordpress.org/ticket/56922. It will hopefully be fixed in WordPress 6.2.

**Comment by georgeh** [comment]
> I have reported this issue upstream here - https://core.trac.wordpress.org/ticket/56922. It will hopefully be fixed in WordPress 6.2.

Thanks @spacedmonkey - I'll get this PR ready for when that gets fixed and hopefully we can backport the fix to this as well.

**Comment by spacedmonkey** [comment]
@georgeh I have pushed some changes to this PR, with a fix. Let me know what you think.

**Comment by georgeh** [comment]
@spacedmonkey thanks, that's about what I was coming up with so it lgtm. How do we feel about merging this when the `_links` section points to URLs that don't work yet?

**Comment by spacedmonkey** [comment]
> @spacedmonkey thanks, that's about what I was coming up with so it lgtm. How do we feel about merging this when the `_links` section points to URLs that don't work yet?

We can fix the issue in core or guternberg PR.

**Comment by spacedmonkey** [comment]
This is looking good. Let's ensure that unit tests / linting are passing. Then we can merge.

**Comment by georgeh** [comment]
@spacedmonkey I'd like to merge this with the version information in `_links` but without the `href`s until the [core ticket](https://core.trac.wordpress.org/ticket/56922) is resolved. I don't know what @TimothyBJacobs availability is to look at that ticket.

```json
{
  "id": "twentytwentythree//home",
  "theme": "twentytwentythree",
  ...
  "_links": {
    ...
    "version-history": [
      {
        "count": 1,
      }
    ],
    "predecessor-version": [
      {
        "id": 17,
      }
    ],
  ...
  }
}
```

This would put the revision data we need in the same place as other post types, but would not advertise URLs that don't work. Once the REST routes are added, we can add the `href`s to the response.

That would allow us to get this feature to users without incurring technical debt for the placement of the data.

**Comment by georgeh** [comment]
After trying this out without the `href`s, it seems that `WP_REST_Server::get_response_links()` really wants those `href`s. I re-added them and cleaned up the warnings. Once we get tests passing can I get a 👍🏻 to merge?

**Comment by georgeh** [comment]
@spacedmonkey All green, can I get a 👍🏻 to merge?

**Comment by georgeh** [comment]
@spacedmonkey I'm going to go ahead and merge tomorrow if I don't hear anything else, my read of your last comment is that we're good to go now that the tests are passing.

File: packages/edit-site/src/store/selectors.js
Comment (dmsnell) [inline packages/edit-site/src/store/selectors.js]: I wish we could avoid bleeding this internal API logic into the store/selectors, but I guess the `getEntityRecord()` API doesn't expose a good way to parse responses or avoid this?

**Review by spacedmonkey — APPROVED** [review]
Looks good to me. This is a merge conflict that @georgeh should have a look into. 

From a PHP standpoint, before this needs up in core, we need to resolve this https://core.trac.wordpress.org/ticket/56922

**Comment by draganescu** [comment]
I've rebased this.

File: lib/experimental/rest-api.php
Comment (draganescu) [inline lib/experimental/rest-api.php:35]: This comment is incorrect. We hook into wp_template_part post type.

File: packages/edit-site/src/store/selectors.js
Comment (draganescu) [reply packages/edit-site/src/store/selectors.js]: I think @dmsnell has a good point on not needing this `getCurrentTemplateRevisionsCount` selector.

Comment (Mamaduka) [reply packages/edit-site/src/store/selectors.js]: +1. We can use `template._links` directly in `useRevisionData` and avoid new selectors.

**Comment by Mamaduka** [comment]
I noticed that the [required fix for the core](https://core.trac.wordpress.org/ticket/56922) has no milestone. So we might have to punt this feature for 6.3.

**Comment by jasmussen** [comment]
Let's ship this, as it's a much needed feature. Thank you George 🙏 

However, let's follow up shortly thereafter with a couple of visual improvements. This is what it looks like at the moment:

Screenshot 2023-02-07 at 09 04 13

That's the cursor hovering over the panel.

The use of a panel component to wrap the button may not have been the best idea, since it appears this panel is nested inside the description. Either the panel should be placed similar to other panels in the document sidebar, go edge to edge and line up with margins and paddings, or it could be a much simpler link, like this:

Screenshot 2023-02-07 at 09 05 17

**Comment by Mamaduka** [comment]
@spacedmonkey, can we merge this feature without https://core.trac.wordpress.org/ticket/56922?

@jasmussen, I agree that Revisions are a needed feature for the Site Editor. However, if the required APIs aren't working correctly, introducing UI doesn't make much sense.

**Comment by jasmussen** [comment]
> However, if the required APIs aren't working correctly, introducing UI doesn't make much sense.

Oh I'll refer to you and others for the call here, I was commenting purely on the placement/feature side of things.

**Comment by TimothyBJacobs** [comment]
I think reasonable people can disagree about this, but I'm fine with merging this as is. The PR doesn't make our existing bug in [#56922](https://core.trac.wordpress.org/ticket/56922) materially worse in my opinion.

**Comment by draganescu** [comment]
Excellent, I'll do the two updates mentioned above and merge!

File: packages/edit-site/src/store/selectors.js
Comment (draganescu) [reply packages/edit-site/src/store/selectors.js]: I removed both new selectors and added the logic in the hook.

File: packages/edit-site/src/store/selectors.js
Comment (Mamaduka) [inline packages/edit-site/src/store/selectors.js]: @draganescu, maybe we should remove this new selector and move logic inside the new hook?

The hooks are private most of the time, and we can easily remove them.

File: packages/edit-site/src/components/sidebar-edit-mode/template-card/last-revision.js
Comment (Mamaduka) [inline packages/edit-site/src/components/sidebar-edit-mode/template-card/last-revision.js]: Suggestion: These values can be fetched/assigned inside `mapSelect`; there's less chance of extra rerenders when comparing the return object with primitive types. What do you think?

**Comment by draganescu** [comment]
@jasmussen I've updated the PR description with the visual update to get it closer to your suggestion.

File: packages/edit-site/src/store/selectors.js
Comment (draganescu) [reply packages/edit-site/src/store/selectors.js]: I think a selector named `getCurrentTemplate` is good! I am less inclined to give this one up :)

File: packages/edit-site/src/components/sidebar-edit-mode/template-card/last-revision.js
Comment (draganescu) [reply packages/edit-site/src/components/sidebar-edit-mode/template-card/last-revision.js]: Good suggestion added in [e82b0d1](https://github.com/WordPress/gutenberg/pull/45215/commits/e82b0d12e5bff3ca277bc21ed85c32d6368d1739)

File: packages/edit-site/src/store/selectors.js
Comment (Mamaduka) [reply packages/edit-site/src/store/selectors.js]: Can we use this selector in other places? There's also the `useEditedEntityRecord` hook that returns similar data.

See: https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/use-edited-entity-record/index.js

Edit: I don't have a strong opinion here.

Comment (draganescu) [reply packages/edit-site/src/store/selectors.js]: Hmm, I didn't know about `useEditedEntityRecord` - I am not a fan of building hooks to avoid obvious selectors, but OTOH let's use what exists for now. Updated in [d6712fd](https://github.com/WordPress/gutenberg/pull/45215/commits/d6712fd8efa3f6e076c5f1f554a2af0af86f7ac9)

Comment (Mamaduka) [reply packages/edit-site/src/store/selectors.js]: The hooks that aren't exported from the packages are easy to modify or remove. But, unfortunately, every selector is part of the public API[^1], and we've to maintain it "forever."

[^1]: Unless marked private via the new package.

Comment (draganescu) [reply packages/edit-site/src/store/selectors.js]: I know about that, and I agree to avoid cluttering the public API with what might be temporary :) needs but usually getters setters for things that will likely never change (templates and template parts for instance) should be added on 1st occasion. What I am saying is `getCurrentTemplate` should have been added when `useEditedEntityRecord` was needed because knowing the currently edited template is a default need in a "template editor" 😁

**Comment by Mamaduka** [comment]
@draganescu, @georgeh, this will require PHP backport. Since this is an enhancement, I believe it must be discussed with the rest of the release squad to get "blessed" status.

Can you create a new ticket on [Trac](https://core.trac.wordpress.org/), so we can start the process?

**Comment by ntsekouras** [comment]
>@draganescu, @georgeh, this will require PHP backport. Since this is an enhancement, I believe it must be discussed with the rest of the release squad to get "blessed" status.

> Can you create a new ticket on [Trac](https://core.trac.wordpress.org/), so we can start the process?

👋 any news on creating a trac ticket for this one? 

FYI: I'll temporarily remove the back port label and re add it to use the `cherry-pick` script.

**Comment by draganescu** [comment]
@Mamaduka @ntsekouras for now I created [Add revisions to the links of the template and template part REST endpoint](https://core.trac.wordpress.org/ticket/57704#ticket) 

I am unclear if the bug in [Template / Template parts revision / autosave REST API are broken](https://core.trac.wordpress.org/ticket/56922) has a mergeable solution in @spacedmonkey 's PR or if we should port over the controller decoration here as a fix for this problem - which we could revert later if the bug is fixed in a more general way (cc @TimothyBJacobs)

**Comment by draganescu** [comment]
This cannot be in 6.2 because once we land in the revision UI the links to take us back don't work and fixing that is not trivial.

**Comment by oandregal** [comment]
I was looking at backporting https://github.com/WordPress/gutenberg/pull/48078 which essentially merges the `Gutenberg_REST_Template_Revision_Count` controller this PR introduced into the existing `Gutenberg_REST_Templates_Controller_6_3`.

I then went to the trac ticket for this PR and saw it was blocked by upstream changes, see at https://core.trac.wordpress.org/ticket/57704#comment:4  It couldn't be backported in 6.3 and, as far as my understanding goes, it cannot be backported in 6.3 either.

As for next steps, fixing the upstream issue would be ideal, and that's the main next step.

I also wonder how we should mark this code as "blocked by upstream". If the upstream issue is not fixed in time for the 6.4 release, I worry that we may unintentionally remove the revisions related code that lives in `lib/compat/6.3`. How can make sure that's obvious, so the people involved in the release have the proper information?

- One idea would be to move the revisions related code to a `lib/compat/6.4`. However, that's pushing the problem to the future, hoping upstream would fix it.
- Alternatively, we could move it to `lib/experimental` and add the proper comments and link to the trac ticket. Code in experimental requires "human intervention", so it's less prone to be unintentionally removed.

How does this sound?

**Comment by ramonjd** [comment]
> Alternatively, we could move it to lib/experimental and add the proper comments and link to the trac ticket. Code in experimental requires "human intervention", so it's less prone to be unintentionally removed.

@oandregal 

Moving files to `/experimental` is what was decided for the Behaviors API:

- https://github.com/WordPress/gutenberg/pull/51654#pullrequestreview-1486482293

To me that makes sense. Then it can be relocated again to `compat/x.y` later when the upstream is fixed.

I can remove the corresponding PR mentions from the [6.3 backport issue](https://github.com/WordPress/gutenberg/issues/51077) if you think we're not going to include it in 6.3?

To confirm, this PR and https://github.com/WordPress/gutenberg/pull/48078 are to be scratched for a future release...

**Comment by ramonjd** [comment]
Here's a PR to move the revisions button API stuff back to `/experimental`

https://github.com/WordPress/gutenberg/pull/51774

**Pull request merged**
